android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java

```diff
@@ -248,103 +237,56 @@ public Dataset fetchDataset(@NonNull Context context, UpdateHandler updateHandle
     int cloudQueryEntries = cloudQueries.size();
     if(cloudQueryEntries > 0) {
       // We need the array to be exactly the same size as our entry count.
-      CloudApiParam[] params = new CloudApiParam[cloudQueryEntries];
+      CloudApiTypes.CloudApiParam[] params = new CloudApiTypes.CloudApiParam[cloudQueryEntries];
       cloudQueries.toArray(params);
-      // We can pass in multiple URLs; this format is extensible if we need extra catalogs in the future.
-      downloadTask.execute(params);
-      executeCallbacks = false;
+      if(USE_DOWNLOAD_MANAGER)
+      {
+        // TODO check duplicated downloads
+        if(CloudDownloadMgr.getInstance().alreadyDownloadingData(DOWNLOAD_IDENTIFIER_CATALOGUE))
+        {
+          String msg = context.getString(R.string.catalog_download_is_running_in_background);
+          Toast.makeText(context, msg, Toast.LENGTH_SHORT).show();
+        }
+        else
+        {
+          String msg = context.getString(R.string.catalog_download_start_in_background);
+          Toast.makeText(context, msg, Toast.LENGTH_SHORT).show();
+          CloudDownloadMgr.getInstance().executeAsDownload(
+            context, DOWNLOAD_IDENTIFIER_CATALOGUE,memCachedDataset,_download_callback,params);
+        }
+
+        executeCallbacks = false;
+      }
+      else {
+        CloudCatalogDownloadTask downloadTask = new CloudCatalogDownloadTask(context, memCachedDataset,_download_callback);
+
+        // We can pass in multiple URLs; this format is extensible if we need extra catalogs in the future.
+        downloadTask.execute(params);
+        executeCallbacks = false;
+      }
     }
 
     // Reuse any valid parts of the cache.
     if(loadKeyboardsFromCache || loadLexicalModelsFromCache) {
-      CloudDownloadReturns jsonData = new CloudDownloadReturns(kbdData, lexData);
+      CloudCatalogDownloadReturns jsonData = new CloudCatalogDownloadReturns(kbdData, lexData);
 
       // Call the processor method directly with the cached API data.
-      downloadTask.processCloudReturns(jsonData, executeCallbacks); // TODO:  Take params for finish, return val for failures
+      _download_callback.processCloudReturns(memCachedDataset,jsonData, executeCallbacks); // TODO:  Take params for finish, return val for failures
     }
 
     return memCachedDataset;
   }
 
-  protected File getKeyboardCacheFile(Context context) {
+  protected static File getKeyboardCacheFile(Context context) {
```

Comment:
Thanks for your contribution on improving this code-base! We really appreciate that you're available to help clean things up in the Android app :)

1. I think it would be good to also move `getKeyboardCacheFile()` and `getLexicalModelCacheFile()` to CloudDataJsonUtil (near `getCachedJSONObject()`) so that all the *Cache.dat accesses go through CloudDataJsonUtil.
2. There's legacy (now redundant) methods in `LanguageListActivity.getCacheFile()` and `LanguageListActivity.getCachedJSONObject()` you should be able to remove.
3. Then you can update whatever referenced "2." to `CloudDataJsonUtil`.